### PR TITLE
Java: IPA the CFG.

### PR DESCRIPTION
**Pull request by aschackmull**
This converts the control flow graph to an IPA type to give us some more options in how we model control flow.  There should be no change in any query results.

### CODE REVIEW

File: java/ql/src/semmle/code/java/ControlFlowGraph.qll [hunk 1]
```diff
@@ -82,44 +82,107 @@
...
-  /** Gets the statement containing this node, if any. */
-  Stmt getEnclosingStmt() {
-    result = this or
-    result = this.(Expr).getEnclosingStmt()
-  }
+module ControlFlow {
+  private newtype TNode =
+    TExprNode(Expr e) { hasControlFlow(e) } or
+    TStmtNode(Stmt s) or
+    TExitNode(Callable c) { exists(c.getBody()) }
```
Comment (hvitved) [inline java/ql/src/semmle/code/java/ControlFlowGraph.qll]: C# also has an entry node, for symmetry.

Comment (hvitved) [reply java/ql/src/semmle/code/java/ControlFlowGraph.qll]: C# also uses a separate IPA type for basic blocks.

Comment (aschackmull) [reply java/ql/src/semmle/code/java/ControlFlowGraph.qll]: I was aiming for minimal changes, so I'm not going to change the CFG with e.g. a new entry node here.  Using a separate type for basic blocks could of course be done, but I don't think it's that important, so I think I'll leave it as is for now.

**Comment by felicity-semmle** [comment]
@aschackmull - I've raised a PR with suggested text changes for the change note.

**Comment by aschackmull** [comment]
> I've raised a PR with suggested text changes for the change note.

Actually, you just added the commit directly to my PR without doing a separate PR against my PR.

**Comment by felicity-semmle** [comment]
Many apologies. I must have managed to click the wrong button - sorry about that.

**Comment by edoardopirovano** [comment]
This might be worth profiling again now that https://github.com/github/semmle-code/pull/39420 is merged to see if that fixed the performance issues.

**Comment by aschackmull** [comment]
It would indeed be interesting to dust off this PR and revisit it, as we have much better tools to deal with performance problems now than back then, but as far as I recall the performance problems weren't related to the many-ipa-branches issue or ipa performance on its own, but rather the problem that ipa types don't have good size estimates, which caused a bunch of cascading changes to various join-orders.

**Comment by aschackmull** [comment]
Finally implemented in https://github.com/github/codeql/pull/17970

**Pull request closed without merging**